django/core/management/commands/dumpdata.py

```diff
@@ -184,7 +218,8 @@ def get_objects(count_only=False):
             if output and self.stdout.isatty() and options['verbosity'] > 0:
                 progress_output = self.stdout
                 object_count = sum(get_objects(count_only=True))
-            stream = open(output, 'w') if output else None
+            open_method, mode = get_open_method(output)
```

Comment:
Instead of this method, can we mirror `loaddata` and store into `self.compression_formats`?

I think it will make things easier to maintain if they're simliar.

We should also store `None: (open, mode),` in that dictionary.

Also what about zip files? You could implement a `SingleZipWriter` to complement `SingleZipReader` in `loaddata`.

I also see that `loaddata` has a special `stdin` case. Do we want a `stdout` here?

Reply:
> Instead of this method, can we mirror `loaddata` and store into self.compression_formats ? I think it will make things easier to maintain if they're similar.

I created a method because I think it's easier to maintain and reuse (e.g. it's imported in the test too)

I think it will be better to have a similar method in `loaddata` too, but I think this PR it's not the right place for this method.

Reply:
> We should also store None: (open, mode), in that dictionary.

The `None` key it's not used, the `(open, mode)` value is returned in the `exception`.

Reply:
> I also see that `loaddata` has a special `stdin` case. Do we want a `stdout` here?

The `stoud` case it's already implemented at line 228 with `stream or self.stdoud`

Reply:
> Also what about zip files? You could implement a `SingleZipWriter` to complement `SingleZipReader` in `loaddata`.

I searched a bit for `zipfile` case but it's not possible to stream into a text file of a zip archive using the stdlib `zipfile`

Similar situation are resolved using third library (e.g. `python-zipstream`) but I don't think we can add a dependencies only to have this feature.

Reply:
So this can be done because we're not trying to stream the output anywhere.

I came up with this which seemed to work well:

```python
import pathlib, zipfile
class SingleZipWriter(zipfile.ZipFile):
    def write(self, data):
        path = pathlib.Path(self.filename)
        name = path.stem if path.suffix == '.zip' else path.name
        return zipfile.ZipFile.writestr(self, name, data)
```

Reply:
I'll test this `SingleZipWriter`.

Reply:
As I tried to explain in the previous comment, zip files are a separate case: the currently supported formats (`gz`, `bz2`, `xz`, `lzma`) are compressed files and the `serialize` method used in` dumpdata` is able to write a text stream to these compressed files and then close them.
The `.zip` format, on the other hand, is a compressed archive (`ZipFile`) that contains compressed files (`ZipInfo`) but the `writestr` method, you suggested, does not continuously append text in the same compressed file inside the zip archive but rather creates a compressed file each time in the archive.
Using the code you suggested, after adapting it to work in the current test, I finally had a `.zip` file containing 69 files compressed each with a part of the fixtures.
I'm still trying to figure out if it's possible to permit to the `serialize` method to `stream` text into `zip` archive without reinventing the wheel, but I would suggest for now to not add support for `.zip` archive in this command, we can simply remove extension from filename and save plain text as it works now.

Reply:
Right, I see. With the following:
```python
f = SingleZipWriter('test.txt.zip', 'w')
f.write('abc')
f.write('123')
f.close()
```
I get:
```
/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'test.txt'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
```
Thanks for taking the time to look.